### ISSUE TITLE
[REVIEW]: UnlockNN: Uncertainty quantification for neural network models of chemical systems

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3700]
**Submitting author:** @a-ws-m (<a href="http://orcid.org/0000-0001-7525-1419">Alexander Moriarty</a>)
**Repository:** https://github.com/a-ws-m/unlockNN/
**Branch with paper.md** (empty if default branch): 
**Version:** v2.0.2
**Editor:** @osorensen
**Reviewers:** @TahiriNadia, @pmeier, @Het-Shah
**Archive:** 10.5281/zenodo.6799685

 

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/b00df538a159c4b6816ec24d4d1716fb/status.svg)](https://joss.theoj.org/papers/b00df538a159c4b6816ec24d4d1716fb)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/b00df538a159c4b6816ec24d4d1716fb"><img src="https://joss.theoj.org/papers/b00df538a159c4b6816ec24d4d1716fb/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/b00df538a159c4b6816ec24d4d1716fb/status.svg)](https://joss.theoj.org/papers/b00df538a159c4b6816ec24d4d1716fb)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/a-ws-m/unlockNN/" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@TahiriNadia & @pmeier & @Het-Shah, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @osorensen know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

## Review checklist for @TahiriNadia
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/a-ws-m/unlockNN/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/a-ws-m/unlockNN/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@a-ws-m) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] #3793
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] #3796
- [x] #3798
- [x] #3799
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] #3797
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @pmeier
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/a-ws-m/unlockNN/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/a-ws-m/unlockNN/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@a-ws-m) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

## Review checklist for @Het-Shah
✨ Important: Please do not use the _Convert to issue_ functionality when working through this checklist, instead, please open any new issues associated with your review <a target="_blank" href="https://github.com/a-ws-m/unlockNN/">in the software repository associated with the submission</a>. ✨

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/a-ws-m/unlockNN/">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@a-ws-m) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @TahiriNadia, @pmeier, @Het-Shah it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 555

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1038/s41586-018-0337-2 is OK
- 10.1021/acs.chemmater.9b01294 is OK
- 10.1038/s41524-020-00406-3 is OK
- 10.1021/acs.jctc.7b00577 is OK
- 10.1557/mrc.2019.78 is OK
- 10.1038/s41524-017-0056-5 is OK
- 10.1038/s41524-019-0221-0 is OK
- 10.1063/1.5019779 is OK
- 10.1088/2632-2153/ab7e1a is OK
- 10.1103/PhysRevLett.120.145301 is OK
- 10.1038/ncomms11241 is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.11 s (387.6 files/s, 37189.4 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          16            408            521           1165
Jupyter Notebook                 1              0            538            379
TeX                              1             16              0            232
Markdown                         4             42              0            160
YAML                             6             15             17            129
reStructuredText                 8             76             55            125
SVG                              1              1              1             53
DOS Batch                        1              8              1             26
INI                              1              2              1             22
JSON                             2              0              0             17
make                             1              4              7              9
-------------------------------------------------------------------------------
SUM:                            42            572           1141           2317
-------------------------------------------------------------------------------


Statistical information for the repository '554da9444b0d05b9d5db864d' was
gathered on 2021/09/07.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Alex Moriarty                  184          6006           2821           52.15
a-ws-m                         155          3504           4595           47.85

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
a-ws-m                     2094           59.8          1.3                9.89
```

**Comment by osorensen** [comment]
@TahiriNadia, @pmeier, @Het-Shah thanks for agreeing to review this submission. This is the review issue, andd you find your review checklist above. Please reach out to me if you have any questions.

**Comment by whedon** [comment]
:wave: @Het-Shah, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @pmeier, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @TahiriNadia, please update us on how your review is going (this is an automated reminder).

**Comment by osorensen** [comment]
@Het-Shah, could you please update us on how it is going with the review? If you have any suggestions for improvement of the software or the paper, feel free to post it here or to open an issue in the [source repository](https://github.com/a-ws-m/unlockNN/).

**Comment by osorensen** [comment]
@pmeier, could you please update us on how it is going with the review? If you have any suggestions for improvement of the software or the paper, feel free to post it here or to open an issue in the [source repository](https://github.com/a-ws-m/unlockNN/).

**Comment by osorensen** [comment]
@TahiriNadia, could you please update us on how it is going with the review? If you have any suggestions for improvement of the software or the paper, feel free to post it here or to open an issue in the [source repository](https://github.com/a-ws-m/unlockNN/).

**Comment by pmeier** [comment]
@osorensen I've checked all the basic things and so far everything looks good. The paper is well written. The only thing left for me is to check its functionality and the documentation thereof. This might take some time, since I'm neither familiar with TensorFlow / keras nor have any domain knowledge. Regarding the latter, I fail to see how I should provide an opinion if the literature references are complete or not.

**Comment by osorensen** [comment]
Thanks for replying, @pmeier! Regarding the literature references, I assume you're referring to this part:

> Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper citation syntax?

I suggest you can interpret this as asking whether papers/datasets/software mentioned inside the paper are properly cited, and whether the citation syntax is correct. So if you don't have an overview of this particular application area, you may still check off the box if the citations are done formally correct.

**Comment by TahiriNadia** [comment]
Sorry, I started to revise the article this weekend, however, I can't do anything, no update!

**Comment by TahiriNadia** [comment]
Sorry, My ckeck list is not clikeable (see below).
![image](https://user-images.githubusercontent.com/19578926/134919516-10e5c060-f8f7-401e-beb1-4cf6bdcf645d.png)

**Comment by danielskatz** [comment]
@whedon re-invite @TahiriNadia as reviewer

**Comment by whedon** [comment]
The reviewer already has a pending invite.

@tahirinadia please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
If that doesn't work (as there's a weird but that sometimes marks the invitation as expired late), let us know and we'll re-issue it - this should let you check the boxes...

**Comment by whedon** [comment]
OK, the reviewer has been re-invited.

@tahirinadia please accept the invite by clicking this link: https://github.com/openjournals/joss-reviews/invitations

**Comment by danielskatz** [comment]
hopefully that one will work

**Comment by TahiriNadia** [comment]
Thank you for your fast response. I confirm it works. I will work today on it.

**Comment by Het-Shah** [comment]
I have opened an issue in the repository for my review.

**Comment by osorensen** [comment]
> I have opened an issue in the repository for my review.

Thanks @Het-Shah. @a-ws-m please place your response in the mentioned issue.

**Comment by osorensen** [comment]
@pmeier, regarding what you wrote yesterday:

>  The only thing left for me is to check its functionality and the documentation thereof.

A general comment, which may or may not apply to this particular case: If you find it hard to test functionality and run the examples, it may well be that other users would experience the same. You might consider whether this is a point at which the software could be improved, and if so, feel free to add a suggestion here or create an issue in the source repository.

**Comment by TahiriNadia** [comment]
I have opened an issue in the repository for my review.

**Comment by pmeier** [comment]
I've added a bunch of issues for my review:

Automated tests and contributing

- a-ws-m/unlockNN#21
- a-ws-m/unlockNN#21
- a-ws-m/unlockNN#21
- a-ws-m/unlockNN#21
- a-ws-m/unlockNN#21
- a-ws-m/unlockNN#21

**Comment by pmeier** [comment]
I've added a bunch of issues about details for my review:

## Automated tests and contributing

- a-ws-m/unlockNN#21
- a-ws-m/unlockNN#22
- a-ws-m/unlockNN#23
- a-ws-m/unlockNN#24
- a-ws-m/unlockNN#25
- a-ws-m/unlockNN#26
- a-ws-m/unlockNN#29
- a-ws-m/unlockNN#30

## Documentation

- a-ws-m/unlockNN#20
- a-ws-m/unlockNN#31
- a-ws-m/unlockNN#32
- a-ws-m/unlockNN#33

---

In general though, I currently agree with the points @Het-Shah brought up in a-ws-m/unlockNN#19. Regarding the substantial scholarly effort, in the 1165 LoC a large amount is actually periphery such as downloading example weights or data or specifically tailored to the MEGNet models. If one removes that, `unlockNN` seems to "only" provide functionality to replace the last layer of a model with one that can also quantify the uncertainty of the prediction as well as metrics to evaluate such a model. I quoted _only_ above, since I'm maybe misreading this. I'm open to a discussion about this to convince me that I'm wrong.

**Comment by pmeier** [comment]
@TahiriNadia 

> I have opened an issue in the repository for my review.

It seems you have opened a bunch of issues in this repository, such as #3793. They were auto-closed by the bot and thus are hardly visible. You probably want to open them in the [repository under review](https://github.com/a-ws-m/unlockNN) instead.

**Comment by osorensen** [comment]
Thanks for noting this, @pmeier. 

@TahiriNadia, I'm sorry about this confusing aspect - you're not the first reviewer who has done this. Could you please open you issues in the source repository instead as @pmeier suggests?

**Comment by osorensen** [comment]
Also, thanks a lot for all the issues you opened in the source repository, @pmeier. @a-ws-m, could you please address these issues. If you expect any of them to take more than a few weeks, please also give us an expected timeline.

**Comment by osorensen** [comment]
Regarding the issue of substantial scholarly effort, as raised by @pmeier and @Het-Shah, the [guidelines](https://joss.readthedocs.io/en/latest/submitting.html) say the following:

> JOSS publishes articles about software that represent substantial scholarly effort on the part of the authors. Your software should be a significant contribution to the available open source software that either enables some new research challenges to be addressed or makes addressing research challenges significantly better (e.g., faster, easier, simpler).
  >
 > As a rule of thumb, JOSS’ minimum allowable contribution should represent not less than three months of work for an 
 individual. Some factors that may be considered by editors and reviewers when judging effort include:
  >
  > - Age of software (is this a well-established software project) / length of commit history.
  > - Number of commits.
  > - Number of authors.
>   - Total lines of code (LOC). Submissions under 1000 LOC will usually be flagged, those under 300 LOC will be desk rejected.
 > - Whether the software has already been cited in academic papers.
 > - Whether the software is sufficiently useful that it is likely to be cited by your peer group.
 > - In addition, JOSS requires that software should be feature-complete (i.e. no half-baked solutions) and designed for maintainable extension (not one-off modifications of existing tools). “Minor utility” packages, including “thin” API clients, and single-function packages are not acceptable.
 
Hence, it is not only an issue of number of lines of code. @pmeier, @Het-Shah, and @a-ws-m, it would be great to hear your thoughts on this. Particularly if more of these points would be satisified if the issues raised by the reviewers are addressed.

**Comment by TahiriNadia** [comment]
> Thanks for noting this, @pmeier.
> 
> @TahiriNadia, I'm sorry about this confusing aspect - you're not the first reviewer who has done this. Could you please open you issues in the source repository instead as @pmeier suggests?

Thanks for the comment @osorensen, as I mentioned I am new to this article evaluation protocol. But I will look at the @pmeier comments as you mention. Thank you!

**Comment by pmeier** [comment]
@osorensen my take on https://github.com/openjournals/joss-reviews/issues/3700#issuecomment-932977470

- :green_circle: **Age of software (is this a well-established software project) / length of commit history**: first commit at July 2020
- :green_circle: **Number of commits**: ~500
- :red_circle: **Number of authors**: `git shortlog -s` reveals two authors, which I think are the same: "Alex Moriarty" and "a-ws-m". This shouldn't be blocking.
- :yellow_circle: **Total lines of code (LOC). Submissions under 1000 LOC will usually be flagged, those under 300 LOC will be desk rejected** see my comments in https://github.com/openjournals/joss-reviews/issues/3700#issuecomment-932956088
- :red_circle: **Whether the software has already been cited in academic papers**: a quick search turned up empty, but this also hard without DOI. This shouldn't be blocking.
- :yellow_circle: **Whether the software is sufficiently useful that it is likely to be cited by your peer group**: hard to judge that without domain knowledge. Let's hope the discussion makes this more clear.
- :green_circle: **In addition, JOSS requires that software should be feature-complete (i.e. no half-baked solutions) and designed for**: package seems complete and flexible enough to not be a one off script

**Comment by osorensen** [comment]
Thanks a lot for your comments, @pmeier!

**Comment by osorensen** [comment]
@a-ws-m, can you please give us an estimate of how long it will take to address the issues raised by the reviewers? Comments to the points are also welcome.

**Comment by a-ws-m** [comment]
Thank you all for the detailed review comments! I'm going to start addressing them this week and I'll hopefully be able to provide an estimate as to how long that'll take over the next few days. Sorry for the delay, I've just been settling into a new postgrad position.

**Comment by osorensen** [comment]
@Het-Shah, could you please check whether your points raised in [this issue](https://github.com/a-ws-m/unlockNN/issues/19) have now been properly addressed by the author, and if so, update your review checklist at the top of this issue?

**Comment by Het-Shah** [comment]
@osorensen I believe @a-ws-m is still working on benchmarks. Once these experiments are done, I will update the checklist.

**Comment by osorensen** [comment]
Thanks for clarifying @Het-Shah, my bad for not being properly updated :-)

**Comment by a-ws-m** [comment]
I am training some models on a benchmark as we type; though they might need some fine-tuning, hopefully that issue will be done by the end of the week.

**Comment by osorensen** [comment]
Thanks for letting us know,  @a-ws-m, and absolutely fine that it takes some time.

**Comment by osorensen** [comment]
@a-ws-m: any news on how it is going with the benchmarks?

**Comment by a-ws-m** [comment]
While I was running the benchmarks I discovered a compatibility issue with the earlier version of TensorFlow Probability I was using. This means I've had to train the models on my local machine, as the cluster's version of CUDA is incompatible with the newer version. The framework is all in place to train the models and I should be able to transfer it over to the new cluster I have just got access to today -- I'll keep you posted.

**Comment by osorensen** [comment]
Thanks @a-ws-m. This is absolutely fine, but if you think you need more than two weeks, I suggest I add the **Paused** label to this issue, so @whedon doesn't give reminders to the reviewers every 14 days. Once you're ready, just let me know in this thread, and we'll start the review again.

**Comment by osorensen** [comment]
I added the pause label, to avoid automatic notifications from `@whedon`, since it seems like this might take some time. We'll remove the label once @a-ws-m is ready.

**Comment by osorensen** [comment]
Any news on this one @a-ws-m?

**Comment by osorensen** [comment]
Thanks @a-ws-m. Regarding @TahiriNadia's comments, note that she has already completed her review checklist, so fixing those issues should not be blocking. However, if you want, you could create those same issues in the source repository yourself, and make sure to include a link to this review issue.

Looking forward to seeing the updates.

**Comment by osorensen** [comment]
@a-ws-m I removed the paused label and look forward to seeing your updates soon.

**Comment by osorensen** [comment]
Dear @a-ws-m, please report some progress within 14th July, or we might have to reject this submission (I also sent this message by e-mail today).

**Comment by Het-Shah** [comment]
@a-ws-m has completed the benchmarks that were required. I have updated and completed the checklist on my end.

**Comment by a-ws-m** [comment]
Thanks @Het-Shah! There are several non-blocking issues on the repository that I addressed a while ago, I'll ask @pmeier to follow up on those separately. I think the only remaining task is RE: substantial scholarly effort.

UnlockNN is specifically tailored to those working within chemistry, hence the focus on applying the uncertainty quantification to ``MEGNet``, which is a successful model for materials property prediction. The bindings needed to make the modifications in this context are quite substantial and required a lot of trial-and-error. Furthermore, for the intended audience, including experimental (rather than computational) researchers, the skills to implement these are less common than for a full-time developer. To that end, the examples are designed to make the package more accessible to them, and the benchmarks are specifically for materials property prediction tasks.

The API component is not limited to applications in chemistry; in principle it should work with any Tensorflow model. Its function is, as [in this comment](https://github.com/openjournals/joss-reviews/issues/3700#issuecomment-932956088), to replace one or more layers of a neural network. I think this is still not trivial due to some quirks of the software packages involved and the method by which the resulting network must be trained, which are accounted for in unlockNN. Although this is now achieved in relatively few lines of code, the [original release](https://github.com/a-ws-m/unlockNN/tree/f90784347cdea83259382fe12885bdf56a72e17f) had ~2400 LoC, which was mostly rewritten; a lot of effort has been invested into streamlining the code.

**Comment by a-ws-m** [comment]
Ah I realise I'd already written a response to the issue about substantial scholarly effort in the thread [here](https://github.com/openjournals/joss-reviews/issues/3700#issuecomment-950791488); I think it's complementary to my last comment. Please let me know if I can provide any more details to address this last task @pmeier.

**Comment by osorensen** [comment]
Thanks a lot @a-ws-m and @Het-Shah!

**Comment by osorensen** [comment]
When you have time, can you please consider whether this addresses the remaining unchecked item on your checklist @pmeier?

**Comment by pmeier** [comment]
@osorensen This review dragged on for a lot longer than I anticipated. I'm currently on parental leave and won't have time to do substantial work for this until approximately mid August.

Seeing that @Het-Shah now accepted the submission with a "Substantial scholarly effort" while he shared similar concerns as I did in the beginning, I would also be fine to just accept it to move forward. Thoughts?

**Comment by osorensen** [comment]
Thanks @pmeier. Given that we've had 3 reviewers, this seems perfectly fine. Could you please check of the last item on your list as well?

Thanks for your efforts in reviewing this submission, and enjoy parental leave!

**Comment by osorensen** [comment]
Thanks @a-ws-m. Once you've merged the changes into the master branch, could you please

- [ ] Make a tagged release of your software, and list the version tag of the archived version here.
- [ ] Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
- [ ] Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
- [ ] Please list the DOI of the archived version here.

Meanwhile I will read through the manuscript once more, and let you know if I have any suggested changes.

**Comment by osorensen** [comment]
@editorialbot generate pdf

**Comment by osorensen** [comment]
@editorialbot check references

**Comment by osorensen** [comment]
> Thanks @a-ws-m. Once you've merged the changes into the master branch, could you please
> 
> * [ ]  Make a tagged release of your software, and list the version tag of the archived version here.
> * [ ]  Archive the reviewed software in Zenodo or a similar service (e.g., figshare, an institutional repository)
> * [ ]  Check the archival deposit (e.g., in Zenodo) has the correct metadata. This includes the title (should match the paper title) and author list (make sure the list is correct and people who only made a small fix are not on it). You may also add the authors' ORCID.
> * [ ]  Please list the DOI of the archived version here.
> 
> Meanwhile I will read through the manuscript once more, and let you know if I have any suggested changes.

@a-ws-m please let me know when this is done, and I can move forward with accepting the submission.

**Comment by a-ws-m** [comment]
Hi, I've made the release: it's [version 2.0.3](https://github.com/a-ws-m/unlockNN/releases/tag/v2.0.3).
The Zenodo DOI is 10.5281/zenodo.6799685.

**Comment by osorensen** [comment]
@editorialbot set version 2.0.3

**Comment by osorensen** [comment]
@editorialbot set 0.5281/zenodo.6799685 as archive

**Comment by osorensen** [comment]
@editorialbot set 10.5281/zenodo.6799685 as archive

**Comment by osorensen** [comment]
@editorialbot generate pdf

**Comment by osorensen** [comment]
@editorialbot check references

**Comment by osorensen** [comment]
Thanks @a-ws-m. Could you please check the latest PR I made in the source repository?

**Comment by a-ws-m** [comment]
Done, thanks for the correction!

**Comment by osorensen** [comment]
@editorialbot generate pdf

**Comment by osorensen** [comment]
@editorialbot check references

**Comment by osorensen** [comment]
@editorialbot recommend-accept

**Comment by danielskatz** [comment]
@editorialbot recommend-accept

thanks, this looks ready to go

**Comment by danielskatz** [comment]
@editorialbot accept

sorry, this is what I meant to do...

**Comment by danielskatz** [comment]
Congratulations to @a-ws-m (Alexander Moriarty) and co-authors!!

And thanks to @TahiriNadia, @pmeier, and @Het-Shah for reviewing, and to @osorensen for editing!
We couldn't do this without you

**Comment by a-ws-m** [comment]
Thank you all for your time and feedback! Feels great to have this published :)
